lib/collection/src/shards/replica_set.rs

```diff
@@ -948,7 +956,43 @@ impl ShardReplicaSet {
         let mut responses = Vec::new();
         let mut errors = Vec::new();
 
-        while let Some(result) = pending_operations.next().await {
+        let mut is_local_operation_resolved = false;
+
+        let update_watcher = async move {
+            match update_watcher {
+                Some(update_watcher) => update_watcher.await,
+                None => future::pending().await,
+            }
+        };
+
+        let update_watcher = update_watcher.fuse();
+
+        tokio::pin!(update_watcher);
```

Comment:
Could you please explain what this do?

Reply:
TL;DR: This makes `update_watcher` future `Unpin`, so that it can be used with `tokio::select`.

<details>
<summary>My simplified explanation of `Pin`/`Unpin`, why do we need them and what `tokio::pin` does</summary>

**NOTE:** All of this is explained in the [`pin`] module documentation.

[`Unpin`] is an auto-trait that simply means "this type is safe to move".

An example of type that is *not* safe to move is any self-referential type.  E.g., imagine the following, but where `reference` points to `object`:

```rust
struct SelfRef {
  object: Object,
  reference: &Object,
}
```

If you move this object, then `reference` would still point to the *previous* address of `object`.

It's generally impossible to create such object in Rust without using `unsafe`... except `async fn` and `async {}` produce self-referential futures. E.g.:

```rust
let future = <some real future>;

async {
  let string = "whatever".to_string();
  let ref = &string;
  
  future.await;
  
  println!("{ref}");
}
```

`string` and `ref` are stored "within" the produced future. And `ref` points to `string`. So it's a self-referential type, which can't be moved. (To be more precise, it can't be moved after first call to `Future::poll` on this future.)

How can we use an unmovable future, then? We have to "pin" it!

There's a [`Pin`] type, that is a wrapper *around a reference*. (The future is unmovable, so once we initialized it, we can't *move* it, we *have to* use a reference.) The idea of `Pin` is that once you pinned a reference (i.e., wrapped reference with `Pin`) the data won't be moved *ever*, even after you drop the `Pin`.

The API of `Pin` is designed to help you not fuck this up: e.g., if your type is `Unpin` (unmovable), you can only create `Pin` using `unsafe`, which should remind you "hey! make sure not to move it *ever*!"

The simplest *practical* way to "pin" a future is `FutureExt::boxed`, which creates `Pin<Box<impl Future>>`. (`Box` implements `Deref`, so it *is* a reference.) This way you can move the `Box` around, but the data would stay unmoved on the heap.

Another way is to use "stack-pinning", which is what `tokio::pin` does. It should do something like that:

```rust
// we have this future in scope
let update_watcher = update_watcher.fuse();

// and here's what `tokio::pin` does
let update_watcher = unsafe { Pin::new_unchecked(&mut update_watcher) };
```

It creates pins the future and forcibly "shadows" original `update_watcher`, so you won't be able to move it once you drop the `Pin`.

And [`Pin<&mut impl Future>` implements `Future`][`pin-future`], but it's also `Unpin`, so we can use it with `tokio::select`.

[`pin`]: https://doc.rust-lang.org/std/pin/index.html
[`Unpin`]: https://doc.rust-lang.org/std/marker/trait.Unpin.html
[`Pin`]: https://doc.rust-lang.org/std/pin/struct.Pin.html
[`pin-future`]: https://doc.rust-lang.org/std/pin/struct.Pin.html#impl-Future-for-Pin%3CP%3E

</details>